[18.0][MIG]contract_sale_generation: Migration to 18.0

javierjcf opened the pull request:
Depends on:

- [x] https://github.com/OCA/contract/pull/1138

Migration from v17 with  changes of this PR #1234 over v17.

angelmoya committed:
ADD contract_sale_generation module

moylop260 committed:
[FIX] contract_payment_auto: Fix freze wkhtmltopdf freezed

angelmoya committed:
[FIX] contract_sale_generation: Isolate sale generation (#118)

Closes #117

pedrobaeza committed:
[FIX] contract_sale_generation: Useless test

pedrobaeza committed:
[10.0][FIX+IMP] contract: Improve usability and don't fail on wrong data (#130)

* [FIX+IMP] contract: Improve usability and don't fail on wrong data

* Cron create invoices masked for avoiding silent errors
* New constraints for assuring data consistency
* UI helps for entering consistent data
* Spanish translation
* Remove double company_id field on form

* [FIX] contract_sale_generation: Adapt tests to upstream contract

flotho committed:
[FIX] contract_sale_generation: it doesn't create sales (#141)

* Change the method called in the view
* Complete the create_invoice method
* Bump version + authoring
* Correct bad call of method
  Small Documentation
* Add super call in python test
* FIX bad field names causing bad quantities in sale.order.line

NL66278 committed:
[FIX] contract. Optimize insertion of dates in invoice and lines.

mvrodriguez committed:
Translated using Weblate (Galician)

Currently translated at 100.0% (10 of 10 strings)

Translation: contract-10.0/contract-10.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-10-0/contract-10-0-contract_sale_generation/gl/

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (10 of 10 strings)

Translation: contract-10.0/contract-10.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-10-0/contract-10-0-contract_sale_generation/de/

jarmokortetjarvi committed:
Translated using Weblate (Finnish)

Currently translated at 100.0% (10 of 10 strings)

Translation: contract-10.0/contract-10.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-10-0/contract-10-0-contract_sale_generation/fi/

murtuzasaleh committed:
[WIP][12.0][MIG] contract_sale_generation

britoederr committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-12.0/contract-12.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-contract_sale_generation/pt_BR/

flotho committed:
add display type in default values

remi-filament committed:
Translated using Weblate (French)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-12.0/contract-12.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-contract_sale_generation/fr/

mkxmawilix committed:
[12.0][FIX] contract: Incorrect product UOM qty dict key for the sale.order.line

The product UOM qty was not correctly set on the Sale Order Line
During the creation of a recurring Order from a Contract, the given quantity on the Contract Line was not correctly set on the created Sale Order Line.

Freni-OSI committed:
[14.0][FIX] Travis

Freni-OSI committed:
[14.0][MIG] contract_sale_generation

rousseldenis committed:
[14.0][IMP] contract_sale_generation

Use the generation_type field defined now in contract base module.

Improve tests

rousseldenis committed:
[IMP] contract_sale_generation: Improve view

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Transmit analytic account

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Add hook method to retrieve sale line values

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Align method on Odoo conventions

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Take into account new method for cron

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Remove deprecated domain onchange

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Split test and common

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Remove unused code

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

rousseldenis committed:
[14.0][FIX] contract_sale_generation: Recompute next recurring date after generation on contract level

As in contract module, recompute recurring_next_date after sale generation on contract level

rousseldenis committed:
[14.0][FIX] contract_sale_generation: Add start date on contract level (for recurrency on contract)

rousseldenis committed:
[14.0][IMP] contract_sale_generation: Add tests with another recurrency

Noel000 committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-14.0/contract-14.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-14-0/contract-14-0-contract_sale_generation/ca/

scigghia committed:
Translated using Weblate (Italian)

Currently translated at 73.3% (11 of 15 strings)

Translation: contract-14.0/contract-14.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-14-0/contract-14-0-contract_sale_generation/it/

ntsirintanis committed:
[MIG] contract_sale_generation: Migration to 15.0

victoralmau committed:
[FIX] contract_sale_generation: Change SavepointCase to TransactionCase

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 93.3% (14 of 15 strings)

Translation: contract-15.0/contract-15.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-15-0/contract-15-0-contract_sale_generation/it/

JesusValdez96 committed:
[MIG] contract_sale_generation: Migration to 16.0

Commit:
[UPD] Update contract_sale_generation.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 83.3% (10 of 12 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (12 of 12 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/es/

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/pt_BR/

aliciagaarzo committed:
[16.0][IMP] contract and contract_sale_generation: Add translations

Commit:
[UPD] Update contract_sale_generation.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/es/

ThijsvOers committed:
Translated using Weblate (Dutch)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/nl/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/it/

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (15 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/pt_BR/

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 86.6% (13 of 15 strings)

Translation: contract-16.0/contract-16.0-contract_sale_generation
Translate-URL: https://translation.odoo-community.org/projects/contract-16-0/contract-16-0-contract_sale_generation/hr/

luisDIXMIT committed:
[FIX] contract_sale_generation: Post-install test + fallback to load CoA

OCA-git-bot committed:
[BOT] post-merge updates

javierjcf committed:
[IMP] contract_sale_generation: pre-commit auto fixes

javierjcf committed:
[MIG] contract_sale_generation: Migration to 17.0

pedrobaeza commented:
It seems this is bringing more changes apart from the module `contract_sale_generation`. Have you done the `git am` properly?

javierjcf commented:
I did it only to take commits of contract_sale_generation from my PR branch:
`git format-patch --keep-subject --stdout origin/18.0..Comunitea/17.0-mig-contract_sale_generation -- contract_sale_generation | git am -3 --keep`

But then i realize i need **contract_sale** dependency on PR #1138... and i did only:
```
git remote add acsone https://github.com/acsone/contract.git
git fetch acsone
git rebase acsone/18.0-mig_contract_sale-sbj
```

I guess I should also run format-patch against the branch from PR #1138, and only apply the commits from the contract_sale module?
Is the order important in this case?

I'm more used to migrating modules that already exist in the official OCA branch.
In this case, since I need to combine two unmerged PRs, I'm not entirely sure about the correct order in which to apply them.

pedrobaeza commented:
For a non merged dependency, check https://github.com/OCA/maintainer-tools/wiki/Use-temporary-reference%28s%29-to-another-pull-request%28s%29

javierjcf commented:
Hi @pedrobaeza 

One question that comes to mind after reviewing the procedure is:

Since this is for v18, my understanding is that by simply editing the pyproject.toml file within the module, I'm telling the test environment that it needs that module, which means this PR can be independent.

What confuses me is having a local branch that is missing the dependency (which it would actually need in order to be installed), but then not committing anything other than the module being migrated.

In summary, my final question is: when a PR has a dependency on one or more other PRs, should the code from those dependencies be included in this PR's branch or not?.

It seems it should be independient, but i'm not sure at all.

pedrobaeza commented:
OCA CI works with pip installable modules, so putting that test_requirements.txt, activate a pip installation over a GitHub pull request, not requiring the files in the traditional way.

javierjcf committed:
[IMP] contract_sale_generation: pre-commit auto fixes

javierjcf committed:
[MIG]contract_sale_generation: Migration to 18.0

javierjcf commented:
Hi @pedrobaeza 

To avoid close the PR i try to fix the branch with: `git reset --hard origin/18.0` and doing the `git am` from my PR #1234 branch.
Then i followed the note about dependencies.

I hope is well done now.

pedrobaeza commented:
Yes, excellent. For a veteran like you, these are peanuts, hehe!

/ocabot migration contract_sale_generation

rousseldenis reviewed (approved):
Code review

AtaponDew reviewed (approved):
It looks good to me, approve.

acsonefho commented on contract_sale_generation/models/contract.py:
```diff
@@ -0,0 +1,132 @@
+        sale = self.env["sale.order"].new(
+            {
+                "partner_id": self.partner_id,
+                "date_order": fields.Date.to_string(date_ref),
```
Not necessary to convert into `str`.

acsonefho commented on contract_sale_generation/models/contract.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        sales = self._recurring_create_sale()
+        for sale_rec in sales:
+            self.message_post(
```
Doesn't work if `self` contains more than one recordset.

acsonefho commented on contract_sale_generation/models/contract.py:
```diff
@@ -0,0 +1,132 @@
+                    sale_values=sale_values,
+                )
+                if invoice_line_values:
+                    sale_values["order_line"].append((0, 0, invoice_line_values))
```
Should be migrated using `Command.xxx(...)`

acsonefho commented on contract_sale_generation/tests/test_contract_sale.py:
```diff
@@ -0,0 +1,148 @@
+        res = {
+            "contract_type": "sale",
+            "contract_line_ids": [
+                (
```
You should also use `Command.xxx(...)` here

javierjcf commented:
Hello,

I applied the corrections from @acsonefho and removed the test-requirements.txt (since contract_sale has already been merged). I did a rebase, and now the failure is happening in the sale_project module — I think because the project is not being propagated correctly.

What’s curious is that on a demo database, with this module and sale_project installed, the tests actually pass.

<img width="1700" height="364" alt="image" src="https://github.com/user-attachments/assets/ddedc37b-9dfb-411f-b06b-b0cc7b9000d3" />

I thought something might be missing because I removed the check for the analytic_account_id field in this module’s test (it no longer exists, I believe it’s now handled with analytic_distribution). But this module doesn’t actually interact with the project or the analytic account…

The fact that the tests pass on a demo database is confusing me.

Any ideas?

NICO-SOLUTIONS commented:
Just wanted to ask whats the status here?
@javierjcf 
After a quick look, it seems that the test test_contract_autoconfirm fails because Odoo’s sale_project module is triggered when confirming the Sale Order. Since the product uses service tracking, sale_project requires a project to create the task – but no project is provided in the test data.

javierjcf commented:
> Just wanted to ask whats the status here? @javierjcf After a quick look, it seems that the test test_contract_autoconfirm fails because Odoo’s sale_project module is triggered when confirming the Sale Order. Since the product uses service tracking, sale_project requires a project to create the task – but no project is provided in the test data.

I'm using this module in production, it works as expected, and in my local enviroment, with sale_project installed (clean db, but no with all contract modules installed) there is no error in the tests...

I dont know if the problem belongs to this module (maybe not doing well with the analytic account/distribution) or is in the oca CI demo data...

NICO-SOLUTIONS commented:
> > Just wanted to ask whats the status here? @javierjcf After a quick look, it seems that the test test_contract_autoconfirm fails because Odoo’s sale_project module is triggered when confirming the Sale Order. Since the product uses service tracking, sale_project requires a project to create the task – but no project is provided in the test data.
> 
> I'm using this module in production, it works as expected, and in my local enviroment, with sale_project installed (clean db, but no with all contract modules installed) there is no error in the tests...
> 
> I dont know if the problem belongs to this module (maybe not doing well with the analytic account/distribution) or is in the oca CI demo data...

@javierjcf 
Haven’t tested it in detail yet... For me, (for now) it looks like the CI tests fail because the demo/test data do not provide a project for the service-tracked product in the test. The issue affects only test/demo data; in production, with full data, it is possible that module works correctly. 
I guess adding a project in the test method or setUp() should fix it, which should be the goal here.

javierjcf commented:
I finally managed to reproduce the error locally. Until now, I was only installing contract_sale_generation and sale_project on a new database (with demo data), and there was no error.
I’m suspecting the sale_timesheet module, because it overwrites the demo product ID that fails:

```xml
<record id="product.product_product_1" model="product.product">
    <field name="service_type">timesheet</field>
    <field name="service_tracking">task_global_project</field>
</record>
```
If i install also sale_timesheet then  i got the same error as in OCA CI

I don’t have much control over the new analytic accounting part (now it’s a widget with JSON format), but I suspect that in the tests I’m not doing anything with it, and maybe I should, since the analytic account (now analytic distribution) used to be linked to the projects.

I’d also like to ask if anyone knows the best way to reproduce OCA’s CI locally… or at least to know which modules are usually installed. It seems like all Odoo addons and all the modules from the given repository. Are the rest of the modules from other repositories included as well?

I also have another question: why do many modules use generic demo data, which may not be installed, instead of declaring their own demo data?

For example, this tests make a mixure of own demo data and odoo demo data

```python
cls.pricelist = cls.env["product.pricelist"].create(
            {"name": "pricelist for contract test"}
        )
        cls.partner = cls.env["res.partner"].create(
            {
                "name": "partner test contract",
                "property_product_pricelist": cls.pricelist.id,
                "email": "demo@demo.com",
            }
        )
        cls.product_1 = cls.env.ref("product.product_product_1")
```

Pricelist and partner are own demo data, but not the product...

In a production database I can’t run these tests, because they require Odoo’s demo data with id=product_product_1. But if the product to test were generated as part of the module’s own test data, and not Odoo’s demo data, then the test could be run on any database. I suppose there’s some reason it’s done this way, but I can’t figure it out.

javierjcf commented:
When testing on a demo database, even after adding the analytic distribution and the group_id field (which would be the analytic account), the error still occurs:

<img width="2513" height="888" alt="image" src="https://github.com/user-attachments/assets/56c5e502-cc00-4937-bbe9-845f247435c3" />

The thing is, in my opinion, a generic demo product (`product.product_product_1`) should not be overwritten

If the sale_timesheet module is not installed, the demo product “Virtual Interior Design” remains normal. But with this module, it’s marked as needing to generate a task, yet no project is associated with it. (You can do it at product level or sale level)

I’m not sure if it’s the responsibility of the demo data in this test to provide a project. If that’s the case, then it should be with the analytic_distribution, right? I tried adding the field to the line, but no luck.

An improvement for me, from my inexperience with tests, would be to use a demo product specific to the module, not Odoo’s generic product_product_1, which several modules overwrite (l10n_in, point_of_sale, sale, and sale_timesheet).

I modified the demo product from `product.product_product_1` to `product.product_product_2`, and as I suspected, the tests now pass. but i'm confused about the use of demo data to run tests.

I don’t have much experience developing tests, so maybe I’m missing something… but wouldn’t it be better to never use generic demo data (since it seems they can be overwritten by other modules, making them inconsistent), and instead have each module be responsible for defining its own demo data?

NICO-SOLUTIONS commented:
I might be wrong, but I think it’s safer not to rely on generic demo data, since other modules can change it. Creating the needed records directly in the test seems more reliable and isolated to me — isn’t that the right approach?

javierjcf commented:
Hi @acsonefho , I did the requested changes, and  solve the test error. I think this is ready to merge?

acsonefho commented:
Thank you @javierjcf for your contribution and this good work! :pray: I hope it'll be merge quickly :muscle:

javierjcf commented:
There was an issue with the analytic account test in v17. In v18, the account_analytic_id field no longer exists in sale.order.

The group_id field set in the contract test data (which represents the analytic account) was lost due to the use of Form() when creating lines. (See discussion in #1234)  However, it can be recovered so that the test—which verifies that generated sales orders carry that analytic account—matches.

In v18, however, the analytic account field was removed from the header of sale order an  the code seems does nothing with this group_id field, so i removed this check from the test in this versión.

rrebollo commented on contract_sale_generation/models/contract_template.py:
```diff
@@ -0,0 +1,15 @@
+    sale_autoconfirm = fields.Boolean()
+
+    generation_type = fields.Selection(
+        selection_add=[("sale", "Sale")],
```
```suggestion
        selection_add=[("sale", "Sale")], ondelete={"sale": "cascade"}
```

etobella replied:
generation_type is not a required field, so ondelete by default is set null :thinking:

rrebollo commented on contract_sale_generation/models/contract.py:
```diff
@@ -0,0 +1,134 @@
+        sales_values = self._prepare_recurring_sales_values(date_ref)
+        sale_orders = self.env["sale.order"].create(sales_values)
+        sale_orders_to_confirm = sale_orders.filtered(
+            lambda sale: sale.contract_auto_confirm
```
```suggestion
            "contract_auto_confirm"
```

etobella reviewed (approved):
LGTM from a technical perspective

@rrebollo Your comments are not related to the migration itself, as this was not changed by @javierjcf and they are not providing a failure. Shall we proceed with merge?

etobella commented:
/ocabot merge patch
